### ISSUE TITLE
ARC GNU 2017.03 - Cannot see symbols in ELF file for assembly files (.s)

### ISSUE THREAD

**Issue by coco21** [foss-for-synopsys-dwc-arc-processors/toolchain#123]
Hello,

We are not able to see symbols related to assembly input source files (.s) in the resulting ELF file.

We use those options shown below in Makefile.

I think it's a simple thing but prefer to ask here directly that would be faster ;)

Thx Christophe

DBG     := -g   -gdwarf-4
# --- Tools --------------------------------------------------------------------
CC := arc-elf32-gcc
ASM := arc-elf32-gcc
STRIP := arc-elf32-strip
LNK := arc-elf32-gcc

# --- File names ---------------------------------------------------------------
EXENAME := $(APP_NAME)
EXE     := $(EXENAME).elf

# Generate mixed assembly and C Listing: ADD This Parameter in the list of files you would like to build:
# Ex: $(CC) $(CCFLAGS) $(GENERATE_LISTING) $(APP_CODE_OPTIM) $(INCLUDES) $${defines} -o $@ $<
ifdef ENABLE_LISTING
 GENERATE_LISTING = -Wa,-adhln=$(<F).lst
else
 GENERATE_LISTING =
endif

# --- Compiler flags -----------------------------------------------------------
CCFLAGS :=           \
  -c                 \
  $(OPTS)            \
  $(DBG)             \
  $(OPTIM)           \
  -Wall              \
  -mlittle-endian    \
  -mcpu=arcem        \
  -mbarrel-shifter   \
  -mmpy-option=0     \
  -mcode-density     \
  -mswap             \
  -mnorm             \
  -fno-jump-tables   \
  -mlong-calls       \
  -save-temps=obj    \
  -mno-sdata \
  -specs=nosys.specs  \
  -Os
  #-mshift-assist

# --- Assembler flags ----------------------------------------------------------
ASMFLAGS :=             \
  -x assembler-with-cpp \
  -c                    \
  $(OPTS)               \
  $(DBG)                \
  $(OPTIM)              \
  -Wall                 \
  -mlittle-endian       \
  -mcpu=arcem           \
  -mbarrel-shifter      \
  -mmpy-option=0        \
  -mcode-density        \
  -mswap                \
  -mnorm                \
  -fmessage-length=0    \
  -mlong-calls          \
  -specs=nosys.specs       \
  -mno-sdata            \
  -Os
  #-mshift-assist
  
# --- Linker flags -------------------------------------------------------------
LNKFLAGS :=               \
  -Wl,-Map,LOCAL/SIMU/BIN/$(EXENAME).map \
  -e __start              \
  $(OPTS)            \
  $(DBG)                  \
  $(OPTIM)                \
  -mmpy-option=0          \
  -mcpu=arcem             \
  -mbarrel-shifter        \
  -mcode-density          \
  -mswap                  \
  -mnorm                  \
  -nostartfiles           \
  -mlong-calls            \
  -specs=nosys.specs      \
  -Wl,--no-check-sections
  #-mshift-assist

**Comment by claziss** [comment]
Are those .s files generated by gcc? or custom ones?

**Comment by coco21** [comment]
This are custom file we wrote and that are given to the toolchain for build.

**Comment by coco21** [comment]
Any idea on that one or do I miss something here? Thx Christophe

**Comment by claziss** [comment]
Maybe the symbols are local or so, are they using [.global](https://sourceware.org/binutils/docs/as/Global.html#Global) directive? Another issue may be strip which can remove symbols. Please can u make a simple example?

If the symbols are functions then please use also [.func](https://sourceware.org/binutils/docs/as/Func.html#Func) directive.

**Comment by coco21** [comment]
We check here further, don't know if we can quickly provide you a (working) example.
What we are missing is line number, absolute file path for custom assembly files, addresses and symbol type seem to be ok (using nm or objdump).
Thx Christophe

**Comment by anthony-kolesov** [comment]
Hi Claudiu,

I can confirm this issue - assembler used to generate line info for assembly files, but then it stopped:

```
$ /u/arcgnu_verif/gnu_builds/2016.09-eng007/elf32_le_linux/bin/arc-elf32-gcc  -g asmloop.S -mcpu=em4 -nostartfiles  -o 2016.09-eng007.elf

$ arc-elf32-objdump -WL 2016.09-eng007.elf

2016.09-eng007.elf:     file format elf32-littlearc

Decoded dump of debug contents of section .debug_line:

CU: asmloop.S:
File name                            Line number    Starting address
asmloop.S                                      4               0x100
asmloop.S                                      5               0x108
asmloop.S                                      6               0x110
asmloop.S                                      7               0x114

$ /u/arcgnu_verif/gnu_builds/2016.09-eng008/elf32_le_linux/bin/arc-elf32-gcc  -g asmloop.S -mcpu=em4 -nostartfiles  -o 2016.09-eng008.elf

$ arc-elf32-objdump -WL 2016.09-eng008.elf

2016.09-eng008.elf:     file format elf32-littlearc

$ cat asmloop.S
.section .text
.global __start
__start:
    mov r0, 0xBEEFDEAD
    mov r1, 0xDEADBEEF
    xor r0, r0, r1
    b __start
```

It seems that it started with 2016.09-eng008, which switched GCC to GCC 6 from 4.8.5 and also "rebased binutils" - I presume you need to compare git tags to see what actual changes were there.

**Comment by coco21** [comment]
Hi Anthony,

Yep timeframes make sense indeed, thanks for this information! I wait then on further steps here, much thanks for your very valuable support!

Christophe

**Comment by anthony-kolesov** [comment]
Hi Chris,

It seems that in GCC 6 the driver doesn't pass option `-g` to the disassembler anymore. You can workaround this by passing `-Wa,--gdwarf2`. 

```
$ arc-elf32-gcc  -g asmloop.S -mcpu=em4 -nostartfiles  -o 2017.09-eng001.elf -Wa,--gdwarf2

$ arc-elf32-objdump -WL 2017.09-eng001.elf

2017.09-eng001.elf:     file format elf32-littlearc

Decoded dump of debug contents of section .debug_line:

CU: asmloop.S:
File name                            Line number    Starting address
asmloop.S                                      4               0x100
asmloop.S                                      5               0x108
asmloop.S                                      6               0x110
asmloop.S                                      7               0x114
```

**Comment by coco21** [comment]
Hi Anthony,

Great, fwd this to teams for test, will confirm back ASAP.
What happens if we want to use DWARF-4?

Thx Chris

**Comment by anthony-kolesov** [comment]
GAS doesn't support dwarf 3 and later, as far as I can see from the binutils/gas/as.c . I guess that DWARF4 doesn't have a lot of effect on debug information for assembly files.

**Comment by coco21** [comment]
Ok I take the information. Seems we tried your approach above already and didn't work. We recheck to be sure we did all right, wait on team feedaback. Chris

**Comment by anthony-kolesov** [comment]
BTW, the option is actually --gdwarf-2, not --gdwarf2, although in my experiments it seems that --gdwarf2 works as well.

**Comment by coco21** [comment]
yep in our case --gdwarf2 works too, we forgot second dash prefix. So it works so far here. Do we have to see this as a life hack or is there a patch planned on your side? Thx Christophe

**Comment by coco21** [comment]
Do you plan to patch something here is has this to be seen as closed in your eyes? Thx Christophe

**Comment by coco21** [comment]
Update from our side since we continue testing this here.

With the flag you propose here (--gdwarf2) we see the code lines in the ELF but sometimes the paths toward the originating assembly file (.s) is wrong. Strangely it sometimes work and sometimes not.

So the approach cannot be seen as working yet since the debugger (ARC GNU 2017.03) cannot find and open the correct assembly files during source level debug sessions.

Any input warmly welcome. Thx Christophe

**Comment by claziss** [comment]
The best is to annotate your custom assembly with debug pseudo ops. Asking the assembler to that for you may lead to inaccurate results. Please, try to use file [.file](https://sourceware.org/binutils/docs/as/File.html#File) it may help you.

**Comment by coco21** [comment]
Update. We found a way that is ok for us for the moment. We include a .h header file into our assembly files and that now works as expected also for source level debug experience.
Coming back to my question above: regarding my (sparse) feedbacks, do you plan a fix on your side to address this or is all functioning as expected and we both are happy here? Thx Christophe

**Comment by claziss** [comment]
Regarding passing -g to the assembler, it seems to be an issue on our side, and I will look into it. There is anything else?

**Comment by claziss** [comment]
The configure script was not checking if ARC assembler can support dwarf2. Fix [in](https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/10f1beb0859f9fe60c94dfc496762e053c4c82fb) :
 ##
[ARC] Check the assembler for gdwar2 support.
gcc/
2017-06-21  Claudiu Zissulescu  <claziss@synopsys.com>

	* configure.ac: Add arc and check if assembler supports gdwar2.
	* configure: Regenerate.

**Comment by claziss** [comment]
I close this issue as there are no comments.
